[Enhancement] formatMessage to accept SafeString arguments

jasonmit opened the pull request:
This is based on a conversation I had with @ericf where he wants to support the use case of passing arguments which are SafeStrings.

```js
Handlebars.registerHelper('strong', function (input) {
  return new Handlebars.SafeString('<strong>' + input + '</strong>');
});
```
```handlebars
{{formatMessage 'Today is {date}' date=(strong (formatDate 'date'))}}
```
Outputs:
Today is **1/20/2014**

```handlebars
{{formatMessage '<em>{date}</em>' date=(strong (formatDate 'date'))}}
```
Outputs (escapes the em tags still):
`&lt;em&gt;`**1/20/2014**`&lt;/em&gt;`

Live example: http://jsbin.com/pekoxipuyo/1/edit?html,output

yahoocla commented:
CLA is valid!

jasonmit commented on src/helpers.js:
```diff
@@ -162,35 +181,20 @@ function registerWith(Handlebars) {
+            message = getMessageFormat(isHtml ? message : escape(message), locales, formats);
+        }
+
+        if (hasSafeString) {
```
If any of the has values is a SafeString, we need to return a SafeString for it to be rendered as HTML.  Everything else is escaped, so shouldn't be a problem.

ericf commented on README.md:
````diff
@@ -367,10 +367,33 @@ The hash arguments represent the name/value pairs that are used to format the `m
 
+```javascript
+Handlebars.registerHelper('strong', function (input) {
+    return new Handlebars.SafeString('<strong>' + input + '</strong>');
````
In this example, we should escape `input`, so if people copy this code, they do it write.

ericf commented on src/helpers.js:
```diff
@@ -162,35 +181,20 @@ function registerWith(Handlebars) {
         // Return a Handlebars `SafeString`. This first unwraps the result to
         // make sure it's not returning a double-wrapped `SafeString`.
-        return new SafeString(String(formatMessage.apply(this, arguments)));
+        return new SafeString(String(formatMessage.call(this, message, options, true)));
```
I generally don't like boolean params, what if instead we cue the `isHTML` behavior on whether the message string itself is an `instanceof SafeString`?

ericf replied:
Also it seems we could end up with a double-wrapped `SafeString` if one of the hash params is a `SafeString`.

ericf commented on src/helpers.js:
```diff
@@ -128,7 +128,7 @@ function registerWith(Handlebars) {
     }
 
-    function formatMessage(message, options) {
+    function formatMessage(message, options, isHtml) {
```
See note about boolean params.

ericf commented on tests/runner.js:
```diff
@@ -10,4 +10,8 @@ global.expect = require('expect.js');
 require('../').registerWith(Handlebars);
 
+Handlebars.registerHelper('strong', function (input) {
+	return new Handlebars.SafeString('<strong>' + input + '</strong>');
```
Escape `input` here.

ericf commented on src/helpers.js:
```diff
@@ -162,35 +181,20 @@ function registerWith(Handlebars) {
+        }
+
+        if (hasSafeString) {
+          return new SafeString(String(message.format(hash)));
```
`message.format()` will always return a native String, so we can remove the `String()` wrapper around this.

ericf commented on src/helpers.js:
```diff
@@ -146,7 +146,26 @@ function registerWith(Handlebars) {
+                value = hash[key];
+
+                if (value instanceof SafeString) {
+                    hash[key] = String(value);
```
What's the reason for getting at the raw string here?

ericf commented:
@jasonmit looking good!

I'm finding the two variables `isHtml` and `hasSafeString` hard to get my head around when thinking of all the possible flows through the code. Can you think of how to possibly simply this so it's easier to follow, or add code comments?

jasonmit commented:
@ericf comments addressed and removed `isHtml`.  I added a comment blurb as to what `hasSafeString` is doing.  Let me know if you're looking for more detail.

Also added a jsbin to demo the functionality. http://jsbin.com/pekoxipuyo/1/edit?html,output

ericf commented on src/helpers.js:
```diff
@@ -144,9 +144,27 @@ function registerWith(Handlebars) {
+                }
+
+                // Escape string value.
+                else if (typeof value === 'string') {
```
I guess when the second `SafeString` is checked it'll be `typeof value === 'object'`.

ericf commented on src/helpers.js:
```diff
@@ -162,35 +180,27 @@ function registerWith(Handlebars) {
 
         if (typeof message === 'string') {
+            if (!(message instanceof SafeString)) {
+                message = escape(message);
```
This code path won't ever be hit if `typeof new SafeString() === 'object'`.

jasonmit replied:
Err, this would always be truthy.  Fixing now.  Essentially need to escape message whenever any of the hash values is a safestring but the message itself was not passed in as a safestring (the changed we talked about this morning to knock out the boolean arg)

ericf commented:
@jasonmit I see you added a unit test already, but could you fill out the unit tests so that we have full branch code coverage in `{{formatMessage}}` and `{{formatHTMLMessage}}` helper functions? I just want to be sure we don't have any security holes since these changes deal with escaping.

ericf commented:
I've thought through this for over an hour and it's really hard to fully understand the all the possible flows through the code (not your fault). This opens up weird behavior around double-escaping when the message is not a string, and also removes the protections that were part of the contract with `{{formatMessage}}`.

I'm going to close this for now, but don't delete the branch. No one has actually been asking for this feature, I thought it would have been a nice to have, but I'm concerned about the added complexity and change in the security contract of these changes.

jasonmit commented:
Sounds good, I'd be happy to revisit any ideas you have around this in the future :+1:
